NIFI-4496: Added JacksonCSVRecordReader to allow choice of CSV parser

mattyb149 opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

andrewmlim commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/csv/CSVReader.java:
```diff
@@ -54,6 +54,26 @@
+            "The CSV parser implementation from the Apache Commons CSV library.");
+
+    public static final AllowableValue JACKSON_CSV = new AllowableValue("jackson-csv", "Jackson CSV",
+            "The CSV parser implementation from the Jackson Dataformats library");
```
Need a period (.) after library to be consistent.

andrewmlim commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/csv/CSVReader.java:
```diff
@@ -54,6 +54,26 @@
+            .name("csv-reader-csv-parser")
+            .displayName("CSV Parser")
+            .description("Specifies which parser to use to read CSV records. NOTE: Different parsers may support different subsets of functionality, "
+                    + "and/or exhibit different levels of performance.")
```
Suggest changing the NOTE to:

Different parsers may support different subsets of functionality and may also exhibit different levels of performance.

jdye64 commented:
@mattyb149 I'm seeing invalid output when I run run an existing flow with this PR. I had an existing flow that used ConvertRecord and Apache Commons CSV. That was working fine and giving me the output I expected. However when I switched to using the Jackson implementation all of the output was empty. I have attached a screenshot from my debugger session in hopes that will help shed some light into what is going on.

<img width="1348" alt="screen shot 2017-11-06 at 4 52 08 pm" src="https://user-images.githubusercontent.com/2127235/32498256-32f8ffc6-c39d-11e7-86dd-cde8f7d3a758.png">

mattyb149 commented:
@jdye64 I think I fixed the issue you were seeing. We have to do most of the schema resolution/management manually, Jackson's methods for handling that don't seem to work for what we need. So I removed the setting of column names on the parser, having the column names changed the parser to want an actual array with [] surrounding the line (weird, right?). Then for files without headers, I needed to make sure we used the schema field names, so I had to adjust the logic where "rawFieldNames" is generated.  Mind taking a look at this latest version? Please and thanks!

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/csv/JacksonCSVRecordReader.java:
```diff
@@ -136,7 +134,7 @@ public Record nextRecord(final boolean coerceTypes, final boolean dropUnknownFie
+                if (!hasHeader || ignoreHeader) {
                     rawFieldNames = schema.getFieldNames();
                 } else {
                     rawFieldNames = Arrays.stream(csvRecord).map((a) -> {
```
I'm not sure that I understand the logic here... was this perhaps due to some refactoring and got overlooked, or is this actually doing something that's just not obvious to me? Seems this could just be done as `Arrays.asList(csvRecord)`

mattyb149 replied:
Who knows lol. I'll try asList() instead

markap14 commented on nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/csv/JacksonCSVRecordReader.java:
```diff
@@ -0,0 +1,257 @@
+            String[] csvRecord = null;
+            try {
+                csvRecord = recordStream.next();
+            } catch (Exception e) {
```
This gives me pause. I think if we catch IOException, we should let it fly. Otherwise, we should wrap the Exception in a MalformedRecordException

mattyb149 replied:
Agreed, that might be left over from a debugging exercise. Definitely not great as-is, will change.
